synapse/handlers/federation.py

```diff
@@ -1695,22 +1695,84 @@ async def on_make_join_request(
             },
         )
 
+        # If the room is a restricted room the auth event ID calculation is a bit
+        # more complicated than the generic methods handle.
+        additional_auth_ids = None
+        current_state_ids = await self.store.get_current_state_ids(room_id)
+        if await self._event_auth_handler.has_restricted_join_rules(
+            current_state_ids, room_version
+        ):
+            additional_auth_ids = [
+                await self._get_user_event_which_could_invite(
+                    room_id, current_state_ids
+                )
+            ]
```

Comment:
This same thing needs to be done for a locally issued join.

Reply:
I'm unsure this is going to work properly, it kind of moves some of the logic about which events to consider in the  auth chain to outside of that calculation (see `auth_types_for_event`). Unfortunately this change means that the events to include in the auth chain depend on the current state of the room (since you need a current membership event in the room to invite).

I'm unsure if this is just some refactoring, or if this breaks an assumption somewhere.